### ISSUE TITLE
[FR] Add media browser

### ISSUE THREAD

**Issue by chemelli74** [elad-bar/ha-shinobi#23]
Would be nice to have media browsing in HA.

Thx for considering,

Simone

**Comment by arag00rn** [comment]
I agree, would be great!

**Comment by elad-bar** [comment]
Added the support, thanks for the FR

**Comment by chemelli74** [comment]
Hi @elad-bar,

thx for the quick addition!
I get the following error while browsing:

![image](https://user-images.githubusercontent.com/57354320/169523465-a998d437-9ab0-4a87-baa7-4ac75c3ecb44.png)

Anything already known ?
Simone

**Comment by elad-bar** [comment]
Will fix it in few minutes, thanks

**Comment by elad-bar** [comment]
@chemelli74 please check with latest version (v1.1.29), if it's working please close the issue.

thanks

**Comment by arag00rn** [comment]
Hi,
I'm using latest version (v1.1.29). I do see videos but I cannot open them.
I tried to debug and seems that it's trying to make a GET on https even if I configured the plugin for standard http.

Thanks,
Andrea

**Comment by chemelli74** [comment]
Hi,

now the list is correctly retrieved, but playback is not proxied.
This means that you need direct access to the shinobi server to stream.
And that you need to configure both Shinobi and HA on the same protocol (http or https).

Not sure if this is the correct info, but hopefully it sheds some light: https://developers.home-assistant.io/docs/core/entity/media-player/?_highlight=proxy#proxy-album-art-for-media-browser

Simone

**Comment by elad-bar** [comment]
That is the media player documentation, not sure it is correct for media source, will check it tomorrow

**Comment by elad-bar** [comment]
Checked it and for Media Browser there is no need for proxy, for media player there is indeed as you have mentioned.

As for http vs https, most casting devices requires transportation to be secured, I have SSL set up for the Shinobi Video as well to allow the streaming of it to any of my TVs and devices

**Comment by chemelli74** [comment]
> Checked it and for Media Browser there is no need for proxy, for media player there is indeed as you have mentioned.

Investigating a bit further, to make it work will all setup ( so make the  streaming via HA ), there is need to work with API views.
One example: 

- https://github.com/blakeblackshear/frigate-hass-integration/blob/1bdaa33c841eba07915503cbdb0083a25f4f3a30/custom_components/frigate/__init__.py#L158-L163

- https://github.com/blakeblackshear/frigate-hass-integration/blob/1bdaa33c841eba07915503cbdb0083a25f4f3a30/custom_components/frigate/views.py

Simone

**Comment by elad-bar** [comment]
It will take time for me to learn and implement the proxy, hope that next week I'll have some good news...

**Comment by elad-bar** [comment]
checked it furter, according to the following thread, media browser handles the proxy by itself:
[home-assistant thread](https://community.home-assistant.io/t/tons-of-questions-about-new-media-browser/229738/37)

will keep checking it, just to understand better the flow I need to check -
HA is configured with SSL, Shinobi without, right?

**Comment by arag00rn** [comment]
Hi,

In my case yes.
HA with SSL, Shinobi without SSL.

Andrea

**Comment by elad-bar** [comment]
What is the error you are getting if you are trying to play with it?

**Comment by arag00rn** [comment]
From Chrome Dev console I see: ERR_SSL_PROTOCOL_ERROR

I guess it's because it's trying to open a http connection using https

**Comment by elad-bar** [comment]
ok, will set it up locally and try to reproduce it
thanks

**Comment by chemelli74** [comment]
HI @elad-bar, any progress ? ;-)

Thx,

Simone

**Comment by chemelli74** [comment]
@elad-bar did you got the time to check this by any chance ?

Thank you in advance,

Simone

**Comment by elad-bar** [comment]
hi @chemelli74,
I didn't manage to reproduce the issue so couldn't solve it,
My HA has 2 addresses for accessing it - http out of the box from HA, https from reverse proxy,
same goes for the Shinobi video.

integration between HA to Shinobi is using the https address of shinobi (to allow streaming / casting for other devices),
to access the HA i'm using https address as well.

Media Browsing is working smoothly from within the house or external network.

if your setup is different, please share so I'll be able to understand how to reproduce that issue.

thanks

**Comment by arag00rn** [comment]
@chemelli74 I solved the issue using a reverse proxy with SSL.

(https) -> HA -> (https) proxy (http) -> Shinobi

**Comment by elad-bar** [comment]
@chemelli74 can you please share the status at your system?

thanks

**Comment by elad-bar** [comment]
@chemelli74 is that issue still relevant?

thanks

**Comment by chemelli74** [comment]
I think so.
My system is:

https(browser) -> reverse(https->http) -> http(HA)

So browser access HA via https but HA access Shinobi via http (and it's not internet exposed).

Simone

**Comment by elad-bar** [comment]
mine is also working the same, https is available only in the reverse proxy, meaning communication between:
- browser to service - will be https -> https
- service to service - can be one of - http -> http / https -> http / https  -> https

not sure i understand what is the issue if that's the case,
if you would like to have proxy because the Shinobi is not exposed to the world that's another motivation (not related to http / https).

thanks

**Comment by chemelli74** [comment]
> if you would like to have proxy because the Shinobi is not exposed to the world that's another motivation (not related to http / https).

If you "proxy" internally in HA the request you get the solution of all scenario:

shinobi -> http -> HA, use proxy view -> https -> browser

And Shinobi is not reachable from internet ;-)

Simone

**Comment by chemelli74** [comment]
Like https://github.com/blakeblackshear/frigate-hass-integration/blob/master/custom_components/frigate/views.py#L227

Simone

**Comment by elad-bar** [comment]
I know how to implement it thanks for the link you have posted above, but i'm not sure about its ROI.

Shinobi Video has several implementations of auth (much stronger than HA), meaning, no real reason that if HA is exposed to the world, Shinobi Video will not.

Will get to it someday but not sure it will be in the upcoming month.

**Comment by elad-bar** [comment]
@chemelli74, just to understand your use case once again, when you are not at home, meaning using the HA externally, do you watch the live stream and the videos from HA?

thanks

**Comment by chemelli74** [comment]
> @chemelli74, just to understand your use case once again, when you are not at home, meaning using the HA externally, do you watch the live stream and the videos from HA?

Yup, would like to see CAM live stream from HA when I'm outside my home. As well as recorded videos or images.

Simone

**Comment by elad-bar** [comment]
Why aren't you exposing the NVR as you are exposing HA to the internet?

thanks

**Comment by chemelli74** [comment]
> Why aren't you exposing the NVR as you are exposing HA to the internet?
> 
> thanks

I'm used to expose the bare minimum ;-)

Everything goes through HA, only MediaBrowser for this integration is not.
So I would love to have it working like it was for Frigate.

Simone

**Comment by elad-bar** [comment]
Just for to explain, the security of HA and Shinobi are pretty mech the same,  (shinobi auth supports also google and ldap auth which ha doesn't),
exposing it through a proxy will extend the delay.

As promised, will try to get it

**Comment by skvalex** [comment]
I would also like to see an option to proxy Shinobi videos to HA as suggested by @chemelli74 as I had to expose the whole Shinobi only to make HA Media Browser work, would prefer to keep it for local network only

**Comment by elad-bar** [comment]
currently pretty busy (work and personal), so not sure I will get to it soon, hopefully in the upcoming month will get to it - with additional feature (support read / unread) and fix (timezone for videos).

if you would like to conribute to the code and suggest PR, feel free.

**Comment by elad-bar** [comment]
support added in v3.0.1, pls check and update

thanks

**Comment by skvalex** [comment]
@elad-bar it works just fine, thank you! Though, I had to re-configure the component (remove and add again).

**Comment by chemelli74** [comment]
> support added in v3.0.1, pls check and update

Which is the best platform to get latest build ? Docker maybe ?

Simone

**Comment by elad-bar** [comment]
> > support added in v3.0.1, pls check and update
> 
> Which is the best platform to get latest build ? Docker maybe ?
> 
> Simone

Not sure I understand the question, there are 2 ways to install it as described in readme
1. Using HACS - recommend 
2. Manual - download zip file and locate the component directory in custom component directory of HA

**Comment by skvalex** [comment]
@elad-bar in the latest version 3.0.1, it seems like the motion event is not being detected by the plugin anymore. Shinobi itself detects it as usual

**Comment by elad-bar** [comment]
For me the sensors are working, pls open another issue (as it's not related to the current feature request) for it with more detail including what version of Shinobi are you using (if docker, which image), you have mentioned plugin, which plugin, logs in debug log at the time of the event

Thanks

**Comment by elad-bar** [comment]
@chemelli74 any updates? 

Thanks

**Comment by chemelli74** [comment]
> @chemelli74 any updates? 
> 
> 
> 
> Thanks

Sorry, I need first to figure out why the new version on shinobi server (just installed) and the latest version of the integration don't generate the camera entity

**Comment by elad-bar** [comment]
Do you have logs in debug level to share?

Thanks

**Comment by chemelli74** [comment]
> Do you have logs in debug level to share?
> 
> Thanks

Sorry was very busy for a house moving.
Here what I see:

```
2023-08-12 19:07:49.378 INFO (MainThread) [custom_components.shinobi.managers.rest_api] Initializing Shinobi Server API
2023-08-12 19:07:49.378 INFO (MainThread) [custom_components.shinobi.managers.rest_api] Status changed from 'Connected to the API' to 'Establishing connection to API'
2023-08-12 19:07:49.449 INFO (MainThread) [custom_components.shinobi.managers.rest_api] Status changed from 'Establishing connection to API' to 'Connected with temporary API key'
2023-08-12 19:07:49.478 INFO (MainThread) [custom_components.shinobi.managers.rest_api] Status changed from 'Connected with temporary API key' to 'Connected to the API'
2023-08-12 19:07:49.517 INFO (MainThread) [custom_components.shinobi.managers.websockets] Status changed from 'Disconnected by the system' to 'Connected to the API'
2023-08-12 19:07:49.517 INFO (MainThread) [custom_components.shinobi.managers.websockets] Starting to listen connected
2023-08-12 19:08:41.375 WARNING (MainThread) [custom_components.shinobi.managers.websockets] Status changed from 'Connected to the API' to 'Not connected'
2023-08-12 19:08:46.378 INFO (MainThread) [custom_components.shinobi.managers.websockets] Status changed from 'Not connected' to 'Disconnected by the system'
```

**Comment by elad-bar** [comment]
Good luck, logs are in info level, when you have time, please share debug log level

Thanks

**Comment by chemelli74** [comment]
> Good luck, logs are in info level, when you have time, please share debug log level

:-(
Too tired probably, sorry.

Enabled now I see the same single line repeated every second:

```
2023-08-12 19:34:21.427 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
2023-08-12 19:34:22.426 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
2023-08-12 19:34:23.425 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
2023-08-12 19:34:24.426 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
2023-08-12 19:34:25.425 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
2023-08-12 19:34:26.426 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
2023-08-12 19:34:27.427 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
2023-08-12 19:34:28.427 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
2023-08-12 19:34:29.425 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
2023-08-12 19:34:30.426 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
2023-08-12 19:34:31.426 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
2023-08-12 19:34:32.425 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
```

Simone

**Comment by elad-bar** [comment]
@chemelli74 I will need the log for the entire loading until you are facing the issue in debug level to understand the issue, for now, in the logs you have shared, I don't see any issue.

thanks

**Comment by chemelli74** [comment]
@elad-bar thx for the patience!

I'm now back home so I could play a bit more. Here the debug full log with a "reload" of the integration to get the full info: 
[home-assistant_shinobi_2023-08-25T17-43-14.686Z.log](https://github.com/elad-bar/ha-shinobi/files/12442484/home-assistant_shinobi_2023-08-25T17-43-14.686Z.log)

Let me know if I can do something more.
If you prefer you can join HomeAssistant  on [Discord](https://discord.com/invite/home-assistant) and then contact me: `chemelli74#2180`

Simone

**Comment by chemelli74** [comment]
BTW, I redacted the username and password in 2 different places:

`2023-08-25 19:39:26.658 INFO (MainThread) [custom_components.shinobi.managers.flow_manager] Config flow started, Step: user, Input: {'title': 'Shinobi Video', 'host': '192.168.1.15', 'path': '/', 'port': 8080, 'ssl': False, 'username': '**redacted**', 'password': '**redacted**plain**'}`

`2023-08-25 19:41:47.697 INFO (MainThread) [custom_components.shinobi.managers.flow_manager] {'host': '192.168.1.15', 'path': '/', 'port': 8080, 'ssl': False, 'username': '**redacted**', 'password': '**redacted**encrypted**', 'title': 'Shinobi Video'}
`

I think you should not log them so it's easier for everyone to share without security concerns.
If you want there is a helper for that in core: https://github.com/home-assistant/core/blob/dev/homeassistant/components/diagnostics/util.py#L25

Simone

**Comment by elad-bar** [comment]
> BTW, I redacted the username and password in 2 different places:
> 
> `2023-08-25 19:39:26.658 INFO (MainThread) [custom_components.shinobi.managers.flow_manager] Config flow started, Step: user, Input: {'title': 'Shinobi Video', 'host': '192.168.1.15', 'path': '/', 'port': 8080, 'ssl': False, 'username': '**redacted**', 'password': '**redacted**plain**'}`
> 
> `2023-08-25 19:41:47.697 INFO (MainThread) [custom_components.shinobi.managers.flow_manager] {'host': '192.168.1.15', 'path': '/', 'port': 8080, 'ssl': False, 'username': '**redacted**', 'password': '**redacted**encrypted**', 'title': 'Shinobi Video'} `
> 
> I think you should not log them so it's easier for everyone to share without security concerns. If you want there is a helper for that in core: https://github.com/home-assistant/core/blob/dev/homeassistant/components/diagnostics/util.py#L25
> 
> Simone

thanks for the feedback and suggestion, applying it to all my component, already available in v3.0.4 (for diagnostic only)

**Comment by chemelli74** [comment]
Any idea why the CAM entity is not created in my setup ?

Simone

**Comment by chemelli74** [comment]
Any idea for investigating?

Simone

**Comment by chemelli74** [comment]
After applying #56 and #57, I have the camera working but not the media browsing:

`2023-10-25 10:33:59.697 ERROR (stream_worker) [homeassistant.components.stream.stream.camera.shinobi_video_terrazzo] Error from stream worker: Stream ended; no additional packets`

![image](https://github.com/elad-bar/ha-shinobi/assets/57354320/8138e8f9-3b04-480f-8876-9834c79bd9d2)

**Comment by elad-bar** [comment]
merging your code and testing it before releasing new version, it doesn't look like related but will check it anyway, thanks for the contribution!

**Comment by chemelli74** [comment]
This is my current status:

```
2023-11-13 18:47:54.670 DEBUG (MainThread) [custom_components.shinobi.media_source] Browse media, Identifier: camera/e3536f7b75cb09effb99bca15d1198e8, Title: Shinobi Video / Shinobi Video
2023-11-13 18:47:54.670 INFO (MainThread) [custom_components.shinobi.media_source] {'e3536f7b75cb09effb99bca15d1198e8': <custom_components.shinobi.managers.coordinator.Coordinator object at 0x7f1d66c52cd0>}
2023-11-13 18:47:54.670 INFO (MainThread) [custom_components.shinobi.media_source] e3536f7b75cb09effb99bca15d1198e8
2023-11-13 18:47:54.670 INFO (MainThread) [custom_components.shinobi.media_source] {'e3536f7b75cb09effb99bca15d1198e8': <custom_components.shinobi.managers.coordinator.Coordinator object at 0x7f1d66c52cd0>}
2023-11-13 18:47:54.670 INFO (MainThread) [custom_components.shinobi.media_source] e3536f7b75cb09effb99bca15d1198e8
2023-11-13 18:47:54.670 DEBUG (MainThread) [custom_components.shinobi.media_source] Building monitors list
2023-11-13 18:47:54.670 DEBUG (MainThread) [custom_components.shinobi.managers.rest_api] GET http://192.168.1.15:8080/KGpA1PAHwOtz1uYAxzKZxK9fNUddA1/videoBrowser/4KKLbBiHNU
2023-11-13 18:47:54.678 DEBUG (MainThread) [custom_components.shinobi.managers.rest_api] Status of http://192.168.1.15:8080/KGpA1PAHwOtz1uYAxzKZxK9fNUddA1/videoBrowser/4KKLbBiHNU: 200
2023-11-13 18:47:54.680 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [139763627365952] Error handling message: Unknown error (unknown_error) admin from <**REDACTED**> (Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36 Edg/119.0.0.0)
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/decorators.py", line 26, in _handle_async_response
    await func(hass, connection, msg)
  File "/usr/src/homeassistant/homeassistant/components/media_source/__init__.py", line 174, in websocket_browse_media
    media = await async_browse_media(hass, msg.get("media_content_id", ""))
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/media_source/__init__.py", line 125, in async_browse_media
    item = await _get_media_item(hass, media_content_id, None).async_browse()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/media_source/models.py", line 79, in async_browse
    return await self.async_media_source().async_browse_media(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/shinobi/media_source.py", line 118, in async_browse_media
    *await action(identifier),
     ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/shinobi/media_source.py", line 221, in _async_build_monitors
    if snapshot.startswith("/"):
       ^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'startswith'
2023-11-13 18:47:56.035 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (s                                                                                                                                                   success: True)
2023-11-13 18:47:56.036 DEBUG (MainThread) [custom_components.shinobi.managers.websockets] Checking event's triggers, Identified: 0
2023-11-13 18:47:57.019 DEBUG (MainThread) [custom_components.shinobi.managers.websockets] Checking event's triggers, Identified: 0
2023-11-13 18:47:57.218 DEBUG (MainThread) [custom_components.shinobi.managers.coordinator] Finished fetching Shinobi Video data in 0.000 seconds (success: True)
```

**Comment by elad-bar** [comment]
I need to apply the same fix you have added in camera to media source component

**Comment by chemelli74** [comment]
> I need to apply the same fix you have added in camera to media source component

Didn't investigate yet, was just to report before loosing the log again ;-)

**Comment by chemelli74** [comment]
Applied the change blindly and we did some progress: I can see the folders and the list of the videos :partying_face:
But still not able to play them:

![image](https://github.com/elad-bar/ha-shinobi/assets/57354320/cfcbb1fa-f9ba-480a-8f79-83dbba8c137b)

**Comment by elad-bar** [comment]
Pls open developer tools in browser and look for error in console

**Comment by chemelli74** [comment]
`Mixed Content: The page at 'https://xxxxxx:8123/media-browser/browser/app%2Cmedia-source%3A%2F%2Fshinobi/app%2Cmedia-source%3A%2F%2Fshinobi%2Fcamera%2Fe3536f7b75cb09effb99bca15d1198e8/album%2Cmedia-source%3A%2F%2Fshinobi%2Fcamera%2Fe3536f7b75cb09effb99bca15d1198e8%2FOTbmfgEU4G10080/album%2Cmedia-source%3A%2F%2Fshinobi%2Fcamera%2Fe3536f7b75cb09effb99bca15d1198e8%2FOTbmfgEU4G10080%2F2023-11-13' was loaded over HTTPS, but requested an insecure video 'http://192.168.1.15:8080/KGpA1PAHwOtz1uYAxzKZxK9fNUddA1/videos/4KKLbBiHNU/OTbmfgEU4G10080/2023-11-13T17-15-01.mp4'. This request has been blocked; the content must be served over HTTPS.`

**Comment by elad-bar** [comment]
As described in the error, you are trying to access unsecured url while the hosting site is secured,
To make it work you need to have shinobi configured as https

**Comment by chemelli74** [comment]
> As described in the error, you are trying to access unsecured url while the hosting site is secured, To make it work you need to have shinobi configured as https

Actually if it's proxied it should not be needed. It was one of the point for asking this implementation ;-)

**Comment by elad-bar** [comment]
Is it configured working with proxy?

**Comment by chemelli74** [comment]
> Is it configured working with proxy?

What you mean exactly ?

**Comment by elad-bar** [comment]
You asked foe the feature and it was released in v3.0.1 as mentioned [above](https://github.com/elad-bar/ha-shinobi/issues/23#issuecomment-1656838892)

Added Use original Stream switch component under device of server

**Comment by elad-bar** [comment]
Sorry, in v3.0.0 - Add proxy view for recordings

**Comment by chemelli74** [comment]
> You asked foe the feature and it was released in v3.0.1 as mentioned [above](https://github.com/elad-bar/ha-shinobi/issues/23#issuecomment-1656838892)
> 
> Added Use original Stream switch component under device of server

Sorry my fault, I though it was autodetected as for other components.
I just enabled "Shinobi Video Server Use Proxy for Recording" switch and now it works fine.

Would you evaluate taking a PR that create the URL from the calling path so it will be always correct ?

Otherwise a user have to change the switch based on the fact that he is or out the home.
